napari/_vispy/overlays/scale_bar.py

```diff
@@ -173,7 +173,35 @@ def _on_box_change(self):
 
     def _on_text_change(self):
         """Update text information"""
-        self.node.text.font_size = self.overlay.font_size
+        # update the dpi scale factor to account for screen dpi
+        # because vispy scales pixel height of text by screen dpi
+        if self.node.text.transforms.dpi:
+            # use 96 as the reference dpi
+            dpi_scale_factor = 96 / self.node.text.transforms.dpi
+        else:
+            dpi_scale_factor = 1
+
+        self.node.text.font_size = self.overlay.font_size * dpi_scale_factor
+        # ensure we recalculate the y_offset from the text size when at top of canvas
+        if 'top' in self.overlay.position:
+            self._on_position_change()
+
+    def _on_position_change(self, event=None):
+        # prevent the text from being cut off by shifting down
+        if 'top' in self.overlay.position:
+            # convert font_size to logical pixels as vispy does
+            # 96 dpi is used as the reference dpi
+            font_logical_pix = self.overlay.font_size * 96 / 72
```

Comment:
adjust for canvas pixel scale

```suggestion
            font_logical_pix = (
                self.overlay.font_size * 96 / 72 *
                self.viewer.window._qt_viewer.canvas._scene_canvas.pixel_scale
            )
```

Reply:
This is not correct. pixel_scale is the conversion between logical pixels and screen pixels. 
See: https://github.com/vispy/vispy/blob/469d337ae24083899e2d5a8f2fa8779de233a568/vispy/app/canvas.py#L357
So to add it here would be doubling things.

Reply:
Is this not resolved? what's the hardcoded 72?

Reply:
@brisvag 72 is the reference vispy uses when computing logical pix for type:
https://github.com/vispy/vispy/blob/af2439895a13f187d840a476bd73ed480f2978f2/vispy/visuals/text/text.py#L588
We need to use that here to ensure the offset accounts for the pixel height of the text.
So 96 is the dpi you used when setting up everything and 72 is the vispy reference dpi.

Reply:
IMO this 96 and 72 reasoning is worth some kind of comment for future contributors! (or a more descriptive name as a constant)

Reply:
I'm open to this. I originally had the link to the vispy code after `as vispy does` but it seemed way too long.

Reply:
yeah at least a comment would be great!

Reply:
I updated the comments. How's it now?

Reply:
Makes sense to me 🎉

Reply:
fwiw I'm a big fan of long comments in code. I once had a look at the CPython dict code and it was like 10 lines of comments per line of code and as a result it was the easiest-to-follow code I have ever read in my life.

Reply:
I meant the link to vispy code is long.

Reply:
That seems even less controversial? 😂